go/vt/throttler/replication_lag_cache.go

```diff
@@ -30,6 +31,8 @@ type replicationLagCache struct {
 	// The map key is replicationLagRecord.LegacyTabletStats.Key.
 	entries map[string]*replicationLagHistory
 
+	mu sync.Mutex
```

Comment:
And for the iteration, we don't need to iterate over the map. We can instead iterate over a slice of the keys. That slice would be our own (a slice is really metadata) with a backing array that is using the same immutable strings that are used for the maps keys:
```
// MaxLag returns the max of all the last replication lag values seen across all tablets of
// the provided type, excluding ignored tablets.
func (t *Throttler) MaxLag(tabletType topodata.TabletType) uint32 {
	cache := t.maxReplicationLagModule.lagCacheByType(tabletType)

	var maxLag uint32

	for _, key := range maps.Keys(cache.entries.Load()) {
```

Reply:
I played around with it for a bit: https://gist.github.com/mattlord/54fcc8de7b98e23292d69258e29d26f7

The mutex may in fact be the way to go — certainly if the perf is adequate for you (I think Slack is the biggest user of the txthrottler today). We could probably use a RWLock instead but that may not make a meaningful difference here either (I'm not even sure how hot this mutex may get). 

Please let me know how your testing goes! If it goes well for you then that's certainly a good signal. I can come back to this quickly at any time. Thanks!

Reply:
@mattlord this would work just as well with `atomic.Pointer`s, yes, the values don't need to be perfectly in sync

I'll make this change because it at least partitions whatever locking `atomic` does behind the scenes 👍

Reply:
> Do we have any idea about perf impact with the mutex?

@mattlord we don't have these numbers. I can write a benchmark to measure but the new/old code but I don't think it's possible to keep the racing code, so I've deferred this for now

Reply:
Agreed with @mattlord re: converting to `atomic.Pointer`. The hint is that the map isn't really updated per se -- just recreated. As such, recreation should swap/replace old map with new map, and reads can run mutex-free once pointer is loaded.

Reply:
@shlomi-noach / @mattlord I partially gave this a shot but I had some concerns/questions here: https://github.com/vitessio/vitess/pull/16078#issuecomment-2388254441

I'm not sure if the mutex is hot enough to see a benefit, but I'm interested in re-attempting `atomic.Pointer` _(or `sync.Map`)_ regardless, they sound pretty cool

Reply:
Ah, I see. Because the throttler is returning the last seen / current value and the mutex is only used when updating that value from the healthcheck results. 👍